Add DMAMUX support

ijager opened the pull request:
**WIP**

This PR will address https://github.com/stm32-rs/stm32g0xx-hal/issues/12. Integrate support for DMAMUX into the DMA driver. 


## Steps to use DMA:


### DMA peripheral init
_Only needs RCC_

- reset peripheral

- enable DMA clock

----


### DMA Channel configure

- Pick DMA channel (1..5, 1..7) (Note: DMAMUX starts at zero)

- Configure DMAMUX  `dmareq_id` with correct index value.

- Configure DMA channel (dma ccr register):

- Write peripheral register in DMA 

- Configure memory buffer address (cparx) and length (cndtrx)

- Enable DMA mode in peripheral
----

### start transfer
- (update buffer addresses)
- enable dma channel (ccr reg -> enable bit)

----


Let's model the DMA driver on [stm32f3xx-hal](https://github.com/stm32-rs/stm32f3xx-hal/blob/master/src/dma.rs). 

Perhaps the DMA Target trait implementations should also configure the DMAMUX, since the peripheral implementation knows which DMAMUX index to set.

```
/// Trait implemented by DMA targets.
pub trait Target {
    /// Enable DMA on the target
    fn enable_dma(&mut self) {}
    /// Disable DMA on the target
    fn disable_dma(&mut self) {}
}
```

andresv commented:
Super!

Yes, I also think that `Target` should configure DMAMUX because index is known.

ijager commented:
I have something working, but it depends on https://github.com/stm32-rs/stm32-rs/pull/408 

I added the `link_dma` method in the `dma::Target` trait

```rust
/// Trait implemented by DMA targets.
pub trait Target {

    /// Connect dmamux channel to dma channel and peripheral
    ///
    /// The implementation should call
    /// dma_ch.select_peripheral(DmaMuxIndex::XXX)
    /// with the correct DmaMuxIndex for XXX
    fn link_dma<C: Channel>(&mut self, dma_ch: &mut C);

    /// Enable DMA on the target
    fn enable_dma(&mut self) {}
    /// Disable DMA on the target
    fn disable_dma(&mut self) {}
}
```

Which needs to be implemented by the peripheral driver that wants to use DMA:

```rust
impl<Config> dma::Target for Rx<$USARTX, Config> {

    fn link_dma<C: dma::Channel>(&mut self, dma_ch: &mut C) {
        dma_ch.select_peripheral(DmaMuxIndex::USART1_RX)
     }
}
```

Example of how to use it:

```rust
    let (mut tx, _rx) = usart1.split();

    let mut dma = dp.DMA.split(&mut rcc, dp.DMAMUX);

    let usart = unsafe { &(*stm32::USART1::ptr()) };
    let tx_data_register_addr = &usart.tdr as *const _ as u32;
    let tx_dma_buf_addr : u32 = tx_buffer.as_ptr() as u32;

    dma.ch1.set_direction(dma::Direction::FromMemory);
    dma.ch1.set_memory_address(tx_dma_buf_addr, true);
    dma.ch1.set_peripheral_address(tx_data_register_addr, false);
    dma.ch1.set_transfer_length(tx_buffer.len() as u16);

    tx.link_dma(&mut dma.ch1);
    tx.enable_dma();
    dma.ch1.enable();
```
What do you think?

andresv commented:
I think it is okay. My only concern is about method `link_dma` name.
Maybe this is more intuitive:
```rust
 tx.use_channel(&mut dma.ch1);
 tx.enable_dma();
 dma.ch1.enable();
```
Or actually it is other way around, we do not want to use channel, channel wants to use peripheral:
```rust
 tx.select_peripheral(&mut dma.ch1);
 tx.enable_dma();
 dma.ch1.enable()
```
But this is also misleading. Basically we would like to configure which peripheral a channel should use, `tx` has other methods so actually yes `_dma` should be in this name. 
```rust
 tx.assign_to_dma(&mut dma.ch1);
// or
 tx.configure_dma(&mut dma.ch1);
 tx.enable_dma();
 dma.ch1.enable();
```

I think this method name should be more clear, but I do not have a good proposal either. Maybe you have some more ideas how to name it?

ijager commented:
perhaps

```rust
tx.use_dma_channel(&mut dma.ch1);
```

andresv commented:
What if:
```rust
dma.ch1.set_transfer_length(tx_buffer.len() as u16);
dma.ch1.select_peripheral(tx.dmamux());
```

ijager commented:
So it is

```rust
tx.use_dma_channel(&mut dma.ch1);
// vs
dma.ch1.select_peripheral(tx.dmamux());
```

I like your suggestion since we're configuring `dma`. Let me try it in the next commit.

andresv commented on examples/uart-dma.rs:
```diff
@@ -0,0 +1,69 @@
+
+    writeln!(usart1, "Hello without DMA\r\n").unwrap();
+
+    let mut tx_buffer = [0u8; 16];
```
You can do this:
```rust
let mut tx_buffer: [u8; 16] = *b"hello with DMA!\n";
```

ijager commented:
I think this can be merged, unless you have there are other issues?
